### ISSUE TITLE
Understanding how functions are traced: example

### ISSUE THREAD

**Issue by lukego** [LuaJIT/LuaJIT#293]
Please indulge my using this issue to think aloud about how LuaJIT traces code. I would like to understand this well enough to explain it to other people, and I think it will help to chat about it here with everybody else who is interested.

The example I am interested in today has these main features:
- One outer loop.
- Three inner loops.
- One `uint64_t` value loaded via FFI and kept live across the inner loops.

I am interested in understanding:
- The order in which the code is traced.
- The way the traces are connected together.
- How the `uint64_t` value is handled (is it boxed? where? why?)

Here is the code:

```
local ffi = require("ffi")
local a = ffi.new('uint64_t[1]')
for i = 1, 1000 do
   x = a[0]               -- x is a 64-bit value loaded from FFI
   for i = 1, 1000 do end -- loop 1
   for i = 1, 1000 do end -- loop 2
   for i = 1, 1000 do end -- loop 3
   z = x                  -- x is still alive here
end
```

### jit.v view

Here is the `jit.v` output:

```
$ ./luajit -jv test2.lua
[TRACE   1 test2.lua:5 loop]
[TRACE   2 test2.lua:6 loop]
[TRACE   3 test2.lua:7 loop]
[TRACE   4 (1/3) test2.lua:6 -> 2]
[TRACE   5 (2/3) test2.lua:7 -> 3]
[TRACE   6 (3/3) test2.lua:8 -> 1]
```

The first three lines tell us that the first inner loop is traced first, then the second inner loop, then the third inner loop. At this point the loops have been compiled to machine code, but at the end of each one the JIT will return to the interpreter before starting the next one.

The other three lines tell us that the JIT then creates traces for three transitions: the first loop (trace 1) to the second loop (trace 2), the second loop to the third loop, and for the third loop back to the first loop. These traces were created because the "exits" at the end of the looping traces were executed often enough to become "hot" and cause a new trace to start where the loop finishes.

The last trace, trace 6, is the most interesting one. This one starts at the end of the third inner loop and then continues back to the first inner loop. This means that it includes all the code in between: both the assignment `z = x` and also the re-entry back into the outer loop. So from the compiler's point of view the start of the outer loop is basically the _last_ thing that happens and the first thing is inner loop number one. The opposite of what I would have imagined :). Funky stuff!

### IR view

Here is how the basic intermediate representation dump looks. I'll break it into three similar pieces. First the inner loops that are all identical:

```
---- TRACE 1 start test2.lua:5
---- TRACE 1 IR
0001    int SLOAD  #7    CI
0002  + int ADD    0001  +1  
0003 >  int LE     0002  +1000
0004 ------ LOOP ------------
0005  + int ADD    0002  +1  
0006 >  int LE     0005  +1000
0007    int PHI    0002  0005
---- TRACE 1 stop -> loop

---- TRACE 2 start test2.lua:6
---- TRACE 2 IR
0001    int SLOAD  #7    CI
0002  + int ADD    0001  +1  
0003 >  int LE     0002  +1000
0004 ------ LOOP ------------
0005  + int ADD    0002  +1  
0006 >  int LE     0005  +1000
0007    int PHI    0002  0005
---- TRACE 2 stop -> loop

---- TRACE 3 start test2.lua:7
---- TRACE 3 IR
0001    int SLOAD  #7    CI
0002  + int ADD    0001  +1  
0003 >  int LE     0002  +1000
0004 ------ LOOP ------------
0005  + int ADD    0002  +1  
0006 >  int LE     0005  +1000
0007    int PHI    0002  0005
---- TRACE 3 stop -> loop
```

Then the transitions from the first loop to the second, and second to third. These traces don't actually do any work and only serve to connect the looping traces together. The IR trace shows that they don't include any instructions:

```
---- TRACE 4 start 1/3 test2.lua:6
---- TRACE 4 IR
---- TRACE 4 stop -> 2

---- TRACE 5 start 2/3 test2.lua:7
---- TRACE 5 IR
---- TRACE 5 stop -> 3
```

Finally the last trace does some real work. Correct me if I am wrong:
- Load the value of `x` because none of the earlier traces did so (it wasn't referenced in the loops.)
- Stores this value in z.
- Updates the outer loop counter and checks for termination.
- Loads the new value of `x` for the start of the next outer loop iteration.

```
---- TRACE 6 start 3/3 test2.lua:8
---- TRACE 6 IR
0001    fun SLOAD  #0    R
0002    tab FLOAD  0001  func.env
0003    int FLOAD  0002  tab.hmask
0004 >  int EQ     0003  +63 
0005    p32 FLOAD  0002  tab.node
0006 >  p32 HREFK  0005  "x"  @33
0007 >  cdt HLOAD  0006
0008 >  p32 HREFK  0005  "z"  @46
0009    tab FLOAD  0002  tab.meta
0010 >  tab EQ     0009  NULL
0011    cdt HSTORE 0008  0007
0012    nil TBAR   0002
0013    num SLOAD  #3    I
0014    num ADD    0013  +1  
0015 >  num LE     0014  +1000
0016 >  cdt SLOAD  #2    T
0017    u16 FLOAD  0016  cdata.ctypeid
0018 >  int EQ     0017  +96 
0019    p64 ADD    0016  +8  
0020    u64 XLOAD  0019  
0021 >  cdt CNEWI  +12   0020
0022    cdt HSTORE 0006  0021
---- TRACE 6 stop -> 1
```

Interesting stuff eh?

### mcode view

So now we have seen which traces are created and how they are connected together. Last question: How is the `uint64_t` value loaded from FFI handled? Is this kept locally in registers, with no heap allocation, or does it get stored on the stack, requiring allocation of a GC-able box?

Looking at IR TRACE 6 above I am specifically wondering about the `CNEWI` instruction at the end that logically allocates a `uint64_t`. Is this "sunk" to avoid a real allocation?

Let's look at the end of the IR trace with `jdump=+rs` enabled to see snapshots, register assignments, and sunk allocations. Here is the relevant bit:

```
---- TRACE 6 IR
...
0021 rax   >  cdt CNEWI  +12   0020
0022          cdt HSTORE 0006  0021
....              SNAP   #3   [ ---- ---- ---- 0014 ---- ---- 0014 +1   +1000 +1   +1   ]
```

Here we can see that the allocation is not sunk. The reason is not that the value is still alive at the end of the trace: if that were so we would see the IR instruction number `0021` referenced in the last line snapshot. (The snapshot is a picture of the Lua stack showing where we need to copy updated values of variables out of registers and back onto the stack.) So what is the reason? I am not sure. Could it be a requirement of the `HSTORE` instruction that the value being stored cannot be sunk? Great if somebody knows and can tell me, I feel like I have misunderstood this part.

Anyway: The machine code shows quite explicitly that this trace does require allowing the GC to run and allocating a new GC-able object:

```
---- TRACE 6 mcode 386
...
f125fd93  call 0x100006bf0	->lj_gc_step_jit
...
f125fe81  call 0x100006f20	->lj_mem_newgco
...
---- TRACE 6 stop -> 1
```

### The end

Does that explanation make any sense? (I wonder if some pictures would help to illustrate the structure of the traces better? Do LOOM and luajit-web-analyzer make this transparent to people already?)

**Comment by corsix** [comment]
> So what is the reason? I am not sure. Could it be a requirement of the `HSTORE` instruction that the value being stored cannot be sunk?

Stores to local variables don't happen in the IR (or the mcode) and are instead recorded in a snapshot and deferred until trace exit. Other kinds of stores can't be recorded in a snapshot, and so have to be recorded in the IR, and so have to happen. As the store has to happen, the cdata allocation has to happen.

**Comment by lukego** [comment]
Thanks for the explanation @corsix.

I had intended to start with `x = a[0]` and end with `y = x` to force the local variable `x` to hold an FFI value throughout the life of the loop. Then it would be interesting to see whether and where the compiler would allocate this. My mental image was a trace starting on the outer loop, loading `x = a[0]` as its first step, and probably allocating the box on the transition into either the first loop (if this were its own trace) or the second loop (if the first were part of the root trace.)

However now I see this is not what happened for at least these reasons:

1. The outer loop was not the start of any trace.
2. I did not declare `x` as a local variable (oops!).
2. The `x = a[0]` load was moved by the compiler to after the loops.
3. The `y = x` assignment is storing the FFI value into a Lua table which, as you pointed out, requires converting it into a garbage collectable heap object.

### Revision

Here is a revised version of the example that's a bit more like what I had in mind: `x` is declared local and its value is stored back in the FFI array. So the allocation could be sunk in principle - we only load and store the 64-bit value from FFI data structures - but it could be very hard for the compiler to keep the value sunk throughout all the traces.

```lua
local ffi = require("ffi")
local a = ffi.new('uint64_t[1]')
for i = 1, 1000 do
   local x = a[0]               -- x is a 64-bit value loaded from FFI
   for i = 1, 1000 do end -- loop 1
   for i = 1, 1000 do end -- loop 2
   for i = 1, 1000 do end -- loop 3
   a[0] = x+1                   -- x is still alive here
end
```

Here is the IR for the last trace where we saw the allocation before:

```
---- TRACE 6 IR
....              SNAP   #0   [ ---- ---- ---- ---- ---- ---- ---- ---- ]
0001 rbp   >  cdt SLOAD  #7    T
0002          u16 FLOAD  0001  cdata.ctypeid
0003       >  int EQ     0002  +12 
0004 rbp      u64 FLOAD  0001  cdata.int64
0005 rbp      u64 ADD    0004  +1  
0006 rax   >  cdt CNEWI  +12   0005
0007 rbx   >  cdt SLOAD  #2    T
0008          u16 FLOAD  0007  cdata.ctypeid
0009       >  int EQ     0008  +96 
0010          p64 ADD    0007  +8  
0011          u64 XSTORE 0010  0005
0012 xmm7     num SLOAD  #3    I
0013 xmm7     num ADD    0012  +1  
....              SNAP   #1   [ ---- ---- ---- ]
0014       >  num LE     0013  +1000
....              SNAP   #2   [ ---- ---- ---- 0013 ---- ---- 0013 0006 +1   +1000 +1   +1   ]
```

Here we see the `CNEWI` allocation at instruction `0006`, which has not been sunk (boxed object is allocated and its address loaded in `rax`), and we see that the reason the allocation did not sink is that the object is live at the end of the trace and needs to be stored on the stack (the `0006` reference in `SNAP #2`.) So in this case there is not such a lot of actual difference in the generated code for the two versions: the 64-bit integer object will be heap allocated whether it is stored on the heap (first version) or the stack (second version.) The main practical difference is that in the first version it's easy to _know_ that the object will be allocated - it's stored in a Lua table and so there is no choice - while in the second version this is ambiguous - how do we determine whether the compiler will be clever enough to sink the allocation?

**Comment by lukego** [comment]
Today I want to look at a new example to explore the limits of allocation sinking in common scenarios. I'll start with a basic example and then test a couple of variations. The theme is to load an FFI value and then inspect it before using it inside a loop. Question is under what conditions will the FFI value needs to be boxed on the heap vs when will allocation sinking kick in.

### Basic example

Here is the first example:

```lua
-- Create an FFI array filled with random numbers
local ffi = require("ffi")
local a = ffi.new('int64_t[10001]')
for i = 1, 10000 do a[i] = 50 + math.random(100) end

for i = 1, 10000 do              -- Loop for each array element
   local n = a[i]                -- Load FFI value into local
   if n > 0 then                 -- Inspect FFI value
      for i = 1, tonumber(n) do  -- Inner loop
      end
   end
end
```

Here is the trace structure:

```
[TRACE   1 test3.lua:3 loop]
[TRACE   2 test3.lua:8 loop]
[TRACE   3 (2/3) test3.lua:5 -> 2]
```

Trace 1 is the loop initializing the random numbers, trace 2 is the inner loop, and trace 3 is from the end of the inner loop back to its start (via re-entry to the outer loop.) This is similar to the previous example. Question of interest is whether we allocate heap objects in trace 2 or trace 3.

The IR for trace 2 (inner loop) has no allocations:

```
---- TRACE 2 start test3.lua:8
---- TRACE 2 IR
....              SNAP   #0   [ ---- ]
0001 rax   >  int SLOAD  #9    CRI
0002       >  int LE     0001  +2147483646
0003 rbp      int SLOAD  #8    CI
0004 rbp    + int ADD    0003  +1  
....              SNAP   #1   [ ---- ---- ---- ---- ---- ---- ---- ---- ]
0005       >  int LE     0004  0001
....              SNAP   #2   [ ---- ---- ---- ---- ---- ---- ---- ---- 0004 0001 ---- 0004 ]
0006 ------------ LOOP ------------
0007 rbp    + int ADD    0004  +1  
....              SNAP   #3   [ ---- ---- ---- ---- ---- ---- ---- ---- ]
0008       >  int LE     0007  0001
0009 rbp      int PHI    0004  0007
---- TRACE 2 stop -> loop
```

There are actually no 64-bit values in this trace anywhere. Presumably the `tonumber(n)` has converted to a number (double) before this trace starts recording.

Trace 3 (exit from inner loop all the way back to start of inner loop) does have an unsunk `CNEWI` instruction that will compile to a heap allocation:

```
---- TRACE 3 start 2/3 test3.lua:5
---- TRACE 3 IR
....              SNAP   #0   [ ---- ---- ---- ---- ---- ---- ---- ---- ]
0001 xmm7     num SLOAD  #3    I
0002 [8]      num ADD    0001  +1  
....              SNAP   #1   [ ---- ---- ---- ]
0003       >  num LE     0002  +10000
....              SNAP   #2   [ ---- ---- ---- 0002 ---- ---- 0002 ]
0004 rbx   >  cdt SLOAD  #2    T
0005          u16 FLOAD  0004  cdata.ctypeid
0006       >  int EQ     0005  +96 
0007 rbp      i64 CONV   0001  i64.num
0008          i64 ADD    0007  +1  
0009          i64 BSHL   0007  +3  
0010          p64 ADD    0009  0004
0011          p64 ADD    0010  +16 
0012 rbp      i64 XLOAD  0011  
0013 rax   >  cdt CNEWI  +11   0012
....              SNAP   #3   [ ---- ---- ---- 0002 ---- ---- ---- ---- ]
0014       >  i64 GT     0012  +0  
....              SNAP   #4   [ ---- ---- ---- 0002 ---- ---- 0002 0013 ]
0015 rbx      fun SLOAD  #0    R
0016 rbx      tab FLOAD  0015  func.env
0017          int FLOAD  0016  tab.hmask
0018       >  int EQ     0017  +63 
0019 rbx      p32 FLOAD  0016  tab.node
0020       >  p32 HREFK  0019  "tonumber" @8
0021       >  fun HLOAD  0020
0022       >  fun EQ     0021  tonumber
0023 xmm5     num CONV   0012  num.i64
0024 rbp   >  int CONV   0023  int.num
0025       >  int LE     0024  +2147483646
....              SNAP   #5   [ ---- ---- ---- 0002 ---- ---- 0002 0013 ]
0026       >  int GE     0024  +1  
....              SNAP   #6   [ ---- ---- ---- 0002 ---- ---- 0002 0013 +1   0023 +1   +1   ]
---- TRACE 3 stop -> 2
```

The allocation is instruction `0013` and we can see that `0013` is referenced in snapshots 3, 4, and 5 so there are multiple execution paths that require the 64-bit value to be copied onto the stack (as a pointer to a boxed object allocated on the heap.) There also seems to be one path of control where the heap object is not needed, but has already been allocated anyway, i.e. if the `n > 0` test fails on instruction `0014` and we restore snapshot 2 which does not reference `0013`.

Generally it is not completely obvious to me how the compiler decides when an object should be allocated and e.g. whether the behaviour is the same for all exits.

### Variation: eliminate the local

Suppose I don't want that allocation to happen. How do I get rid of it? There is actually a fairly easy kludge: don't assign the value to a local variable. Here is a revision that references the value inside the array `a[i]` instead of putting it into a variable with `local n = a[i]`:

```lua
-- Create an FFI array filled with random numbers
local ffi = require("ffi")
local a = ffi.new('int64_t[10001]')
for i = 1, 10000 do a[i] = 50 + math.random(100) end

for i = 1, 10000 do                 -- Loop for each array element
   if a[i] > 0 then                 -- Inspect FFI value
      for i = 1, tonumber(a[i]) do  -- Inner loop
      end
   end
end
```

This variation compiles almost the same except that allocation sinking eliminates the heap object allocation. Here is the IR for trace 3:

```
---- TRACE 3 start 2/3 test4.lua:6
---- TRACE 3 IR
....              SNAP   #0   [ ---- ---- ---- ---- ---- ---- ---- ]
0001 xmm5     num SLOAD  #3    I
0002 xmm7     num ADD    0001  +1  
....              SNAP   #1   [ ---- ---- ---- ]
0003       >  num LE     0002  +10000
....              SNAP   #2   [ ---- ---- ---- 0002 ---- ---- 0002 ]
0004 rbx   >  cdt SLOAD  #2    T
0005          u16 FLOAD  0004  cdata.ctypeid
0006       >  int EQ     0005  +96 
0007 rbp      i64 CONV   0001  i64.num
0008          i64 ADD    0007  +1  
0009          i64 BSHL   0007  +3  
0010          p64 ADD    0009  0004
0011          p64 ADD    0010  +16 
0012 rbp      i64 XLOAD  0011  
0013  {sink}  cdt CNEWI  +11   0012
....              SNAP   #3   [ ---- ---- ---- 0002 ---- ---- ---- ]
0014       >  i64 GT     0012  +0  
....              SNAP   #4   [ ---- ---- ---- 0002 ---- ---- 0002 ]
0015 rbx      fun SLOAD  #0    R
0016 rbx      tab FLOAD  0015  func.env
0017          int FLOAD  0016  tab.hmask
0018       >  int EQ     0017  +63 
0019 rbx      p32 FLOAD  0016  tab.node
0020       >  p32 HREFK  0019  "tonumber" @8
0021       >  fun HLOAD  0020
0022       >  fun EQ     0021  tonumber
0023 xmm5     num CONV   0012  num.i64
0024 rbp   >  int CONV   0023  int.num
0025       >  int LE     0024  +2147483646
....              SNAP   #5   [ ---- ---- ---- 0002 ---- ---- 0002 ]
0026       >  int GE     0024  +1  
....              SNAP   #6   [ ---- ---- ---- 0002 ---- ---- 0002 +1   0023 +1   +1   ]
---- TRACE 3 stop -> 2
```

See that the `CNEWI` allocation at instruction `0013` has been eliminated (trace shows `{sink}` as the location). This is possible because none of the snapshots reference `0013` i.e. we never need to store the 64-bit value on the stack. This means that we can access the 64-bit value directly in register `rbp` where it was loaded by the `XLOAD` at `0012`.

Note also that although we wrote `a[i]` twice in the source code the value was only actually loaded once because the compiler can see that we didn't do anything that could affect the value in between the two loads e.g. an FFI call that might poke the array.

### The end

So: looks like allocation sinking is not effective if we have objects like 64-bit integers or FFI pointers stored in local variables and we are using simple control constructs like loops. The extreme kludge of avoiding assigning the values to variables has succeeded on this example. It remains to be seen if there is some middle ground e.g. ways that we can limit the scope or usage of our locals to be confident that they will not require allocations.

Question is: How should a programmer write their code in cases where they want to avoid allocations? Do they avoid assigning FFI values to locals? Or limit the scope of those locals in some specific way e.g. make sure they are not referenced - or not lexically available - where there are loops and branches?

Tips would be welcome :-). And more experiments needed.

**Comment by javierguerragiraldez** [comment]
hi Luke, very intersting tests!  on the latter "don't use a variable" to avoid an allocation, i'd guess this only helps within an inner loop, if the FFI value has an escape from the trace, then it must be store somewhere, either a Lua variable, or at least in a snapshot.

As a teaser, here are your Lua examples with the HTML produced by loom.  Please tell if they're more or less readable than `-jdump` output.  (And issues/patches on loom's repo are welcome, of course!)
[traces.zip](https://github.com/LuaJIT/LuaJIT/files/816344/traces.zip)

**Comment by lukego** [comment]
Just some passing wild ideas:

How about if we allowed storing immutable 64-bit values directly on the stack? Then we could convert this to a heap object only if we need to store a reference on the heap e.g. storing the value in a Lua table or a closure. If this were only for immutable values, e.g. FFI pointers and int64, then we would not have to worry about referencing both the stack and the heap representation of the same value because they are both interchangeable.

I have observed that allocations correlate with performance overhead but I have not teased out the details yet. How much overhead is the object creation (`lj_mem_newgco()`), how much is giving the GC a chance to wake up (`lj_gc_step_jit()`), and how much is when the GC needs to do serious work? Could be possible to optimize some of these cases... for example if we are concerned about cases where we know that we will create a modest amount of heap objects and they they will nearly all become garbage at a certain point in time, then we might be able to defer the collection until the end, both skipping the `lj_gc_step_jit()` calls on each allocation (do this all at once later) and potentially making life easy for a generational collector to sweep away a nursery generation. Some more study needed here too...

**Comment by lukego** [comment]
Thanks @javierguerragiraldez!

**Comment by lukego** [comment]
@javierguerragiraldez The LOOM view is really nice for these examples. I like having the overview of the source code and showing how this maps onto bytecode and onto traces. Bonus points for showing in gray where a trace is repeating the outer loop!

![screen shot 2017-03-03 at 10 40 42](https://cloud.githubusercontent.com/assets/13791/23546068/40158cfa-fffe-11e6-9902-900d5c3e5995.png)

The mouseover cross referencing is handy too e.g. putting the mouse over the IR instruction for the `CNEWI` instruction and seeing where it is referenced (in this case by the last snapshot):

![screen shot 2017-03-03 at 10 45 31](https://cloud.githubusercontent.com/assets/13791/23546172/ab60c0f6-fffe-11e6-8147-0087214e7ad9.png)

Have you thought about the "does it scale" question already? I find that in practice my jit.dump is for a real application usually around 200,000 lines long with hundreds of traces. This presents a few problems for me. One is finding the needle in the haystack i.e. which trace should I actually care about. This is why I have been writing trace-oriented profilers. One more problem is actually being able to produce the trace efficiently: the current `jit.dump` module has too much performance overhead and enabling it during benchmarking interferes with the results. So I need a more streamlined version of `jit.dump` that defers the CPU-heavy work of decoding traces until after runtime. Can you relate?

**Comment by lukego** [comment]
> on the latter "don't use a variable" to avoid an allocation, i'd guess this only helps within an inner loop, if the FFI value has an escape from the trace, then it must be store somewhere, either a Lua variable, or at least in a snapshot.

I am avoiding this situation in my example by referring to the FFI value as `a[i]` i.e. referencing it inside an FFI array each time. My current mental model says that this way if two traces need to access the value then they will each load it from FFI with a separate `XLOAD` instead of passing it to each other via the stack. (In principle this would make every access load the value separately but some optimization - CSE? - seems to eliminate that redundancy when there is no BARRIER between the references.)

This seems to work in practice for me but it feels awkward and also becomes verbose if I am not writing `a[i]` but repeating a longer reference like `mystructure.elements[i].data` every time I want to reference the value.

**Comment by lukego** [comment]
@javierguerragiraldez The examples over at https://github.com/LuaJIT/LuaJIT/issues/208#issuecomment-236423732 could make interesting test cases for LOOM? There you have e.g. multiple traces for the same line of code (side traces) and also multiple prototypes for the same source code. Just if you are looking for some potential edge cases :).

**Comment by lukego** [comment]
(Closing to reduce clutter.)
